pkg/cloud/ratelimit.go

```diff
@@ -112,3 +112,108 @@ func (m *MinimumRateLimiter) Accept(ctx context.Context, key *RateLimitKey) erro
 func (m *MinimumRateLimiter) Observe(ctx context.Context, err error, key *RateLimitKey) {
 	m.RateLimiter.Observe(ctx, err, key)
 }
+
+type TickerRateLimiter struct {
+	ticker *time.Ticker
+}
+
+func NewTickerRateLimiter(limit int, interval time.Duration) *TickerRateLimiter {
+	return &TickerRateLimiter{
+		ticker: time.NewTicker(interval / time.Duration(limit)),
+	}
+}
+
+// Accept everything immediately.
+func (t *TickerRateLimiter) Accept(ctx context.Context, rlk *RateLimitKey) error {
+	select {
+	case <-t.ticker.C:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Observe does nothing.
+func (*TickerRateLimiter) Observe(context.Context, error, *RateLimitKey) {
+}
+
+// CompositeRateLimiter combines rate limiters based on RateLimitKey.
+type CompositeRateLimiter struct {
+	rateLimiters map[string]map[string]map[string]RateLimiter
+	defaultRL    RateLimiter
+}
+
+// NewCompositeRateLimiter creates a new CompositeRateLimiter that will use
+// provided default rate limiter if no better match is found.
+func NewCompositeRateLimiter(defaultRL RateLimiter) *CompositeRateLimiter {
+	m := map[string]map[string]map[string]RateLimiter{
+		"": {
+			"": {
+				"": defaultRL,
+			},
+		},
+	}
+	return &CompositeRateLimiter{
+		rateLimiters: m,
+		defaultRL:    defaultRL,
+	}
+}
+
+func (c *CompositeRateLimiter) ensureExists(project, service string) {
+	if _, ok := c.rateLimiters[project]; !ok {
+		c.rateLimiters[project] = map[string]map[string]RateLimiter{}
+	}
+	if _, ok := c.rateLimiters[project][service]; !ok {
+		c.rateLimiters[project][service] = map[string]RateLimiter{}
+	}
+}
+
+func (c *CompositeRateLimiter) fillMissing() {
+	for _, subProject := range c.rateLimiters {
+		if subProject[""] == nil {
+			subProject[""] = map[string]RateLimiter{}
+		}
+		for _, subService := range subProject {
+			if subService[""] == nil {
+				subService[""] = c.defaultRL
+			}
+		}
+	}
+}
+
+// Register adds provided rl to the composite rate limiter.
+//
+// It replaces previous rate limiter provided for the same project, service,
+// operation combination. Once a rate limiter is added, it can't be removed.
+//
+// Same rate limiter can be used for multiple Register calls.
```

Comment:
It seems like there is a special semantic on matching when you pass in an empty string "". Can we document this contract in the comments?